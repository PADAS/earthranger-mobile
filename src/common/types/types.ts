import { EventStatus, PatrolResult } from '../enums/enums';

export type Position = [number, number];

export type RootStackParamList = {
  SplashScreen: undefined,
  LoginView: undefined,
  TrackLocationMapView: {
    flyTo?: Position;
  }
  MainTabBar: undefined,
  ReportsView: undefined,
  ReportIssueView: undefined,
  PinAuthenticationView: undefined,
  SyncLoaderView: undefined,
  ReportCategoriesView: {
    coordinates?: Position,
  },
  ReportTypesView: {
    title: string,
    categoryId: string,
    coordinates?: Position,
    isPatrolInfoEventType?: boolean,
  },
  ReportForm: {
    reportId?: number,
    title: string,
    typeId: string,
    coordinates?: Position
    schema: string,
    geometryType: string,
    isEditMode?: boolean,
    createdAt?: string,
    isDefaultPatrolInfoEnabled?: boolean,
    geoJson?: GeoJSON.FeatureCollection,
    areaMeters?: number,
    perimeterMeters?: number,
    polygonPoints?: string,
    viewBox?: string,
    isPatrolInfoEvent?: boolean,
  },
  ReportEditLocationView: { coordinates: Position },
  RecordReportAreaView: undefined,
  AboutView: undefined,
  ReportNoteView: { id: number, text: string, coordinates: Position, channelId: string },
  MenuSettingsView: undefined,
  SettingsView: undefined,
  StatusView: undefined,
  ReportRepeatableFieldListView: {
    schema: any,
    title: string,
    rendererId: string,
    data?: RepeatableFieldListData[],
  },
  RepeatableFormView: {
    schema: any,
    title: string,
    listId: string,
  },
  EventsListFilterView: undefined,
  ReportsPendingSync: undefined,
  TrackedBySubjectsView: {
    subjects?: any,
  },
  BasemapView: undefined,
  CoordinateUnitsView: undefined,
  PatrolsTypeView: {
    coordinates: Position,
  },
  StartPatrolView: {
    patrolTitle: string,
    patrolType: string,
    patrolTypeId: string,
    coordinates: Position,
  },
  PhotoQualityView: undefined,
  ResetDatabaseCacheView: undefined,
  SubjectsListView: {
    parentSubjectGroupName?: string,
    isParentView?: boolean,
    parentId?: string,
  }
};

export type BottomSheetParamList = {
  Basemap: undefined,
  PatrolDetailsView: { patrolId: number },
  PatrolTrack: undefined,
  TrackingStatusView: undefined,
  SubjectDetailsView: undefined,
};

export interface EdgeInsets {
  top: number;
  right: number;
  bottom: number;
  left: number;
}

export type RepeatableFieldListData = {
  id: number;
  title: string;
  data: any;
  hidden: boolean,
};

export type LastLocation = {
  timestamp: string;
  displayDate: string;
  latitude: number;
  longitude: number;
  accuracy: number;
};

export interface IconProps {
  width?: string;
  height?: string;
  color?: string;
  viewbox?: string;
  isDisabled?: boolean;
}

export type AttachmentPhotoType = 'photo';
export type AttachmentNoteType = 'note';
export type AttachmentUpdateNoteType = 'updateNote';
export type AttachmentAddNewNoteType = 'addNewNote';
export type AttachmentDeleteNoteType = 'deleteNote';

export type ReportFormEvent = {
  id: number;
  remoteId: string,
  accountId: string,
  eventTypeId: number;
  title: string;
  latitude: number;
  longitude: number;
  geometry: string;
  event_values: string;
  isDraft: number;
  createdAt: number,
  updatedAt: number,
  patrol_segment_id: number | undefined,
};

export type EventData = {
  eventTypeId: string;
  title: string;
  latitude: number;
  longitude: number;
  geometry: string;
  event_values: string;
  isDraft: number;
  createdAt: string,
  updatedAt: string,
  patrol_segment_id: number | undefined,
};

export type ReportFormFullEvent = {
  account_id: number;
  profile_id: number;
  event_type_id: string;
  title: string;
  latitude: number;
  longitude: number;
  geometry: string,
  event_values: string;
  is_draft: number;
  value: string;
  updated_at: string;
  created_at: string;
  id: string;
  geometry_type: string;
  segment_remote_id: string | undefined;
  patrol_segment_id: string | undefined;
};

export type PersistedEventType = {
  id?: number;
  remote_id: string;
  account_id: number;
  profile_id: string;
  value: string;
  display: string;
  schema: string;
  category_id: string;
  geometry_type: string;
  default_priority: string;
  icon: string;
  icon_svg: string;
};

export type PersistedPatrolType = {
  id: number;
  remote_id: string;
  account_id: number;
  value: string;
  display: string;
  icon: string;
  icon_svg: string;
  default_priority: number;
  is_active: number;
  is_selected: boolean;
};

export type PersistedPatrol = {
  id: number;
  remote_id: string;
  account_id: number;
  profile_id: number;
  title: string;
  priority: string;
  state: string;
  created_at: string;
  updated_at: string;
};

export type PersistedPatrolSegment = {
  id: number
  remote_id: string;
  patrol_id: number;
  patrol_type_id: number;
  start_latitude: string;
  start_longitude: string;
  stop_latitude: string;
  stop_longitude: string;
  start_time: string;
  end_time: string;
};

export type PersistedPatrolTypeSegment = {
  id: number;
  remote_id: string;
  patrol_id: number;
  patrol_type_id: number;
  start_latitude: string;
  start_longitude: string;
  stop_latitude: string;
  stop_longitude: string;
  start_time: string;
  end_time: string;
  patrol_type: string;
};

export type DraftReports = {
  id: number;
  title: string;
  type_id: number,
  updated_at: string;
  icon_svg: string;
  default_priority: number;
  schema: string,
  geometry_type: string,
  hidden?: boolean;
  is_draft?: boolean;
};

export type EventUploadDetails = {
  reportId: number;
  reportTitle: string;
  isReportUploaded: boolean;
  uploaded: {
    notes: number;
    images: number;
  };
  pending: {
    notes: number;
    images: number;
  };
  state: string;
};

export type ReportFormEventAttachments = {
  id: number,
  remote_id: string,
  profile_id: number,
  account_id: number,
  event_id: string;
  type: string;
  path: string;
  thumbnail_path: string,
  note_text: string;
  uploaded: number,
  attachment_id: string;
};

export type PersistedPatrolDetail = {
  remote_id: string,
  title: string,
  serial_number: string,
  start_latitude: number,
  start_longitude: number,
  start_time: number,
  icon_svg: number,
  created_at: number,
};

export type PatrolDetail = {
  remoteId: string,
  title: string,
  serialNumber: string,
  startLatitude: number,
  startLongitude: number,
  startTime: number,
  iconSVG: string
  startTimeFormatted: string,
  createdAt: number,
};

export type PersistedUserProfile = {
  id: number;
  remote_id: string;
  account_id: string;
  username: string;
  content_type: string;
  subject_id: string;
  pin: string;
  permissions: string,
};

export type UserInfo = {
  id: number,
  type: string,
  permissions: string,
};

export type PersistedUserSubject = {
  id: number;
  remote_id: string;
  account_id: string;
  name: string;
  content_type: string;
  isSelected: boolean;
};

export type Patrol = {
  id: number,
  typeId: string,
  type: string,
  display: string,
  coordinates: Position | null,
  status: PatrolResult,
  timestamp: string,
};

// Polygon Reducer

export type PolygonState = {
  past: Position[],
  current: Position[],
};

export type PolygonAction = {
  type: 'UNDO' | 'REDO' | 'ADD',
  value?: Position,
};

type TGeometry = 'polygon' | 'point';

export const Geometry: Record<TGeometry, string> = {
  polygon: 'polygon',
  point: 'point',
};

export type Size = {
  width: number,
  height: number,
};

export interface EventListItem {
  bgColor: string,
  defaultPriority: number,
  error: string,
  fgColor: string,
  hidden: boolean,
  icon: string,
  id: number,
  isDraft: boolean,
  labelText: string,
  remoteId: string,
  status: EventStatus,
  statusIcon: string,
  text: string,
  title: string,
  errorMsg?: string,
  hasErrorOnEvent: boolean,
}

export type SubjectGroupData = {
  count: number,
  isVisible: boolean,
  title: string,
  type: 'group',
  subjects: number[],
  id: string,
};

export type Subject = {
  id: number,
  name: string,
  type: 'subject',
  isVisible: boolean,
  icon: string,
  lastPosition: Position,
  lastPositionUpdate: string,
  isHidden: boolean,
};

export type SubjectGroupSyncItem = {
  name: string,
  remoteId: string,
  parentId?: string,
  parentLocalId?: string,
};

export type SubjectSyncItem = {
  iconSVG: string,
  isActive: string,
  lastPosition: string,
  lastPositionDate: string,
  name: string,
  remoteId: string,
  tracksAvailable: string,
  updatedAt: string,
};
