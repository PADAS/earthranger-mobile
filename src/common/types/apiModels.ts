export interface FileRequest {
  uri: string,
  type: string,
  name: string,
}

export enum ApiStatus {
  Unknown = 1,
  Succeeded = 200,
  BadRequest = 400,
  Unauthorized = 401,
  Forbidden = 403,
  NotFound = 404,
  ServerError = 500,
}
