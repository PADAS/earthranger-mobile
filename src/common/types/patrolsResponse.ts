import { Subject } from './usersResponse';

export interface Status {
  code: number;
  message: string;
}

export interface PatrolType {
  default_priority: number;
  display: string;
  icon_id: string;
  id: string;
  is_active: boolean;
  value: string;
  ordernum: number;
}

export interface PatrolTypesResponse {
  data: PatrolType[];
  status: Status;
}

export interface TrackPatrolByResponse {
  data: {
    properties: {
      leader: {
        enum: Subject [],
      }
    }
  }
  status: Status,
}
