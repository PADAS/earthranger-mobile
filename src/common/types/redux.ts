export interface InitialState {
  session: {
    accessToken: string,
    refreshToken: string,
    error: null,
    sessionError: null,
  },
  reports: {
    noteId: number,
    textNote: string,
  }
}

export type ActionType = 'REFRESH_TOKEN'
| 'REFRESH_TOKEN_ERROR'
| 'USER_LOGIN'
| 'USER_LOGIN_ERROR'
| 'GET_SESSION'
| 'GET_SESSION_ERROR'
| 'CREATE_NOTE'
| 'CLEAR_NOTE_DATA'
| 'NUMBER_OF_REPORT_DRAFTS'
| 'CLEAR_SESSION';

export interface ReducerAction {
  type: ActionType,
  payload: any;
}
