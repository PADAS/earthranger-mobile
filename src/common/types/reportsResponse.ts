export interface ReportCategoriesResponse {
  data: Category[];
  status: Status;
}

export interface Category {
  id: string;
  value: string;
  display: string;
  ordernum: number;
  flag?: string;
  account_id: string | undefined;
  profile_id: number[];
}

export interface ReportTypesResponse {
  data: Type[];
  status: Status;
}

export interface ReportTypeSchemaResponse {
  data: string;
  status: Status;
}

export interface Type {
  id: string;
  icon: string;
  value: string;
  display: string;
  ordernum: number;
  is_collection: boolean;
  readonly?: boolean;
  category: Category;
  icon_id: string;
  is_active: boolean;
  default_priority: number;
  geometry_type: string;
  schema: string;
  url: string;
  icon_svg: string;
}

export interface Status {
  code: number;
  message: string;
}

export interface EventCategory {
  id: number;
  remote_id: string;
  value: string;
  display: string;
}

export interface EventType {
  id: number;
  display: string;
  icon_svg: string;
  default_priority: string;
}

export interface Note {
  id: number,
  text: string,
}

export interface WithChildren {
  children: any;
}

export interface UploadedReport {
  id: number;
}
