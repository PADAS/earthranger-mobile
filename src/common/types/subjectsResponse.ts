export interface SubjectGroupResponse {
  data: SubjectGroup[],
  status: Status,
}

export interface SubjectGroup {
  name: string,
  id: string,
  subgroups: SubjectGroup[],
  subjects: Subject[],
}

export interface Subject {
  content_type: string,
  id: string,
  name: string,
  subject_type: string,
  subject_subtype: string,
  common_name: string | null,
  additional: SubjectAdditional,
  created_at: string,
  updated_at: string,
  is_active: boolean,
  user: null,
  region: string,
  country: string,
  sex: string,
  tracks_available: boolean,
  image_url: string,
  last_position_status: SubjectLastPositionStatus,
  last_position_date: string,
  last_position: SubjectLastPosition,
  device_status_properties: SubjectDeviceStatusProperties[],
  url: string,
}

interface SubjectAdditional {
  rgb: string,
  sex: string,
  region: string,
  country: string,
  external_id: string,
  tm_animal_id: string,
  external_name: string,
}

interface SubjectLastPositionStatus {
  last_voice_call_start_at: string | null,
  radio_state_at: string | null,
  radio_state: string,
}

interface SubjectLastPosition {
  type: string,
  geometry: {
    type: string,
    coordinates: [number, number]
  },
  properties: {
    title: string,
    subject_type: string,
    subject_subtype: string,
    id: string,
    stroke: string,
    stroke_opacity: number,
    stroke_width: number,
    image: string,
    last_voice_call_start_at: string | null,
    location_requested_at: string | null,
    radio_state_at: string,
    radio_state: string,
    coordinateProperties: {
      time: string,
    },
    DateTime: string,
  }
}

interface SubjectDeviceStatusProperties {
  label: string,
  units: string,
  value: number,
}

interface Status {
  code: number;
  message: string;
}
