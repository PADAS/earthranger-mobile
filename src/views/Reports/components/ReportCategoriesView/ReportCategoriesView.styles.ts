// External Dependencies
import { StyleSheet } from 'react-native';

// Internal Dependencies
import { COLORS_LIGHT } from '../../../../common/constants/colors';

const styles = StyleSheet.create({
  content: {
    flex: 1,
    backgroundColor: COLORS_LIGHT.white,
  },
  listRow: {
    borderBottomWidth: 1,
    borderColor: COLORS_LIGHT.G5_LightGreyLines,
  },
  listRowText: {
    color: COLORS_LIGHT.G0_black,
    fontSize: 16,
    marginBottom: 20,
    marginLeft: 16,
    marginTop: 20,
    marginRight: 16,
  },
});

export default styles;
