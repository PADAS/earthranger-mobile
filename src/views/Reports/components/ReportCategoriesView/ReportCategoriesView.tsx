// External Dependencies
import React, { useEffect, useState } from 'react';
import {
  View,
  Text,
  FlatList,
  TouchableOpacity,
} from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { NativeStackNavigationProp, NativeStackScreenProps } from '@react-navigation/native-stack';

// Internal Dependencies
import { useRetrieveReportCategories } from '../../../../common/data/reports/useRetrieveReportCategories';
import { EventCategory } from '../../../../common/types/reportsResponse';
import { RootStackParamList, Position } from '../../../../common/types/types';
import { PermissionView } from '../../../Permission/PermissionView';
import { UserType } from '../../../../common/enums/enums';
import { useRetrieveUser } from '../../../../common/data/users/useRetrieveUser';

// Styles
import styles from './ReportCategoriesView.styles';

interface ReportCategoriesViewProps {
  navigation: NativeStackNavigationProp<RootStackParamList, 'ReportCategoriesView'>;
  route: NativeStackScreenProps<RootStackParamList, 'ReportCategoriesView'>;
}

const ReportCategoriesView = ({
  route,
  navigation,
}: ReportCategoriesViewProps) => {
  // Hooks
  const { retrieveUserInfo } = useRetrieveUser();
  const { retrieveReportCategoriesByUserType } = useRetrieveReportCategories();

  // Component's State
  const [reportTypes, setReportTypes] = useState<EventCategory[]>([]);
  const [coordinates, setCoordinates] = useState<Position>();

  useEffect(() => {
    getReportCategories();
    initMapCoordinates();
  }, []);

  // Handlers
  const getReportCategories = async () => {
    const userInfo = await retrieveUserInfo();
    if (userInfo?.userType) {
      const profileId = userInfo?.userType === UserType.profile ? parseInt(userInfo.userId || '', 10) : undefined;
      const reportCategoriesList = await retrieveReportCategoriesByUserType(
        userInfo.userType,
        profileId,
      );
      setReportTypes(reportCategoriesList);
    }
  };

  const initMapCoordinates = () => {
    if (route.params.coordinates) {
      setCoordinates(
        [
          parseFloat(route.params.coordinates[0].toFixed(6)),
          parseFloat(route.params.coordinates[1].toFixed(6)),
        ],
      );
    }
  };

  const onPress = (categoryId: string, title: string) => {
    navigation.navigate('ReportTypesView', {
      categoryId,
      coordinates,
      title,
    });
  };

  return (
    <SafeAreaView style={styles.content} edges={['bottom']}>
      {(reportTypes.length > 0) ? (
        <View>
          {/* List of Report Types */}
          <FlatList
            data={reportTypes}
            renderItem={({ item }) => (
              <TouchableOpacity onPress={() => {
                onPress(item.remote_id, item.display);
              }}
              >
                <View style={styles.listRow} key={item.id}>
                  <Text style={styles.listRowText}>{item.display}</Text>
                </View>
              </TouchableOpacity>
            )}
          />
          {/* End List of Report Types */}
        </View>
      ) : (<PermissionView />)}
    </SafeAreaView>
  );
};

export { ReportCategoriesView };
