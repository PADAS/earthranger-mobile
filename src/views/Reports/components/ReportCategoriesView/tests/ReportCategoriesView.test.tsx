// External Dependencies
import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';

// Internal Dependencies
import { ReportCategoriesView } from '../ReportCategoriesView';
import { RootStackParamList } from '../../../../../common/types/types';

// Types + Interfaces
// eslint-disable-next-line max-len
type StackNavigationPropAlias = NativeStackNavigationProp<RootStackParamList, 'ReportCategoriesView'>;

// Test Specific Mocks
jest.mock('../../../../../common/data/reports/useRetrieveReportCategories', () => ({
  useRetrieveReportCategories: () => ({
    retrieveReportCategoriesByUserType: () => Promise.resolve([{
      id: 1,
      display: 'Test',
    }]),
  }),
}));

describe('ReportCategoriesView', () => {
  let navigation: Pick<StackNavigationPropAlias, 'navigate'>;
  const route = {
    params: {
      coordinates: [1, 2],
    },
  };

  beforeEach(() => {
    navigation = {
      navigate: jest.fn(),
    };
  });

  it('matches snapshot', async () => {
    const { toJSON, getByText } = render(
      // @ts-ignore
      <ReportCategoriesView navigation={navigation} route={route} />,
    );

    await waitFor(() => {
      expect(getByText('Test')).toBeTruthy();
    });

    expect(toJSON()).toMatchSnapshot();
  });
});
