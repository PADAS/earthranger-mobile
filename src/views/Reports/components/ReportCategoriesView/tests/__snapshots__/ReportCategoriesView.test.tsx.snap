// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportCategoriesView matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <RCTScrollView
      data={
        Array [
          Object {
            "display": "Test",
            "id": 1,
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "borderBottomWidth": 1,
                  "borderColor": "#DDDDDD",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#000000",
                    "fontSize": 16,
                    "marginBottom": 20,
                    "marginLeft": 16,
                    "marginRight": 16,
                    "marginTop": 20,
                  }
                }
              >
                Test
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaView>
`;
